File: Javascript/day03/day-03-b.js
const fs = require("fs");

const text = fs.readFileSync("./input.txt", "utf8").trimEnd();

let totalPriority = 0;

let elves = text.split("\n");

for (let i = 0; i < elves.length; i += 3) {
  let group = elves.slice(i, i + 3);
  let badge = findBadge(group);
  totalPriority += prioritize(badge);
}

console.log(totalPriority);

/**
 *
 * @param {string} item
 * @returns {number}
 */
function prioritize(item) {
  if (item >= "a") {
    return item.charCodeAt(0) - "a".charCodeAt(0) + 1;
  } else {
    return item.charCodeAt(0) - "A".charCodeAt(0) + 27;
  }
}

/**
 *
 * @param {string[]} group
 * @returns {string}
 */
function findBadge(group) {
  return group[0]
    .split("")
    .find((item) => group[1].includes(item) && group[2].includes(item));
}

File: Go/day03/day-03-b.go
package main

import (
	"fmt"
	"os"
	"strings"
	"unicode"
)

func main() {
	bytes, _ := os.ReadFile("./input.txt")
	text := string(bytes)

	elves := strings.Split(text, "\n")
	totalPriority := 0

	for i := 0; i < len(elves); i+=3 {
		group := elves[i: i+3]
		badge := findBadge(group)
		totalPriority += prioritize(badge)
	}

	fmt.Println(totalPriority)
}

func prioritize(item rune) int {
	if unicode.IsLower(item) {
		return int(item) -'a' + 1
	} else {
		return int(item) - 'A' +27
	}
}

func findBadge(group []string) rune {
	common := 'a'

	for _, item := range group[0] {
		if strings.ContainsRune(group[1], item) && strings.ContainsRune(group[2], item) {
			common = item
		}
	}

	return common
}